src/main/java/org/opentripplanner/graph_builder/module/osm/WayPropertySet.java

```diff
@@ -70,11 +81,112 @@ public WayPropertySet() {
+        Map<String, String> wayTags = way.getTags();
+        if (wayTags == null) {
+            // no tags added to way. Probably occurs only during testing.
+            return "";
+        }
+        for (Entry<String, String> wayTagValue : wayTags.entrySet()) {
+            String tag = wayTagValue.getKey();
+            String value = wayTagValue.getValue();
+            String tagValue = String.format("%s=%s", tag, value);
+
+            // create list of tags to see if any of them are in applicable list
+            List<String> possibleMatches = new ArrayList<>();
+
+            // add exact match use case
+            possibleMatches.add(tagValue);
+
+            // add wildcard value for tag
+            possibleMatches.add(String.format("%s=*", tag));
+
+            // if value contains a colon and thus possibly a suffix (such as surface=cobblestone:flattened), use the
+            // root tag value
+            String rootValue = null;
+            if (value.contains(":")) {
+                rootValue = value.split(":", 2)[0];
+                // add exact tag wth root value
+                possibleMatches.add(String.format("%s=%s", tag, rootValue));
+            }
+
+            // if tag contains a colon and thus a suffix (such as cycleway:right), use the root tag key to check if
+            // there is a match
+            if (tag.contains(":")) {
+                String rootTag = tag.split(":", 2)[0];
+                // add both exact and wilcard tag/values
+                possibleMatches.add(String.format("%s=%s", rootTag, value));
+                possibleMatches.add(String.format("%s=*", rootTag));
+
+                // also add rootValue if applicable
+                if (rootValue != null) {
+                    possibleMatches.add(String.format("%s=%s", rootTag, rootValue));
+                }
+            }
+
+            for (String possibleMatch : possibleMatches) {
+                if (possibleTagValues.contains(possibleMatch)) {
+                    // add the exact tag value to make sure cache key is for this exact set of tag/values
+                    applicableTagValues.add(tagValue);
+                    // break as there is no need to search for other possibilities
+                    break;
+                }
+            }
+        }
+        // sort values to make deterministic key
+        sort(applicableTagValues);
+        return String.join(";", applicableTagValues);
+    }
+
     /**
      * Applies the WayProperties whose OSMPicker best matches this way. In addition, WayProperties that are mixins
      * will have their safety values applied if they match at all.
      */
     public WayProperties getDataForWay(OSMWithTags way) {
+        String wayPropertiesKey = getKeyFromApplicableTags(way, possibleWayPropertyTagValues);
+        // check if lookup key found
+        WayProperties cachedWayProperties = wayPropertyLookup.get(wayPropertiesKey);
+        if (cachedWayProperties != null) {
```

Comment:
Potential for using `HashMap.computeIfAbsent(key, mapper)` (Java 8) https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#computeIfAbsent-K-java.util.function.Function-